File: churn_reduction.R
rm(list = ls())
setwd("E:/1ST SEM/eng/edwisor_assignments/8.project 2")
## Read the data
df_train = read.csv("Train_data.csv", header = T)
df_test = read.csv("Test_data.csv", header = T)
#structure of data
str(df_train)
#Load Libraries
x = c("ggplot2", "corrgram", "DMwR", "caret", "randomForest", "unbalanced", "C50", "dummies", "e1071", "Information",
      "MASS", "rpart", "gbm", "ROSE", 'sampling', 'DataCombine', 'inTrees')
lapply(x, require, character.only = TRUE)
rm(x)
#Missing value analysis
missing_train = data.frame(apply(df_train,2,function(x){sum(is.na(x))}))
missing_test = data.frame(apply(df_test,2,function(x){sum(is.na(x))}))



df=df_train
#d1 = df
numeric_index = sapply(df_train,is.numeric) #selecting only numeric
numerical = df_train[,numeric_index]
Numerical = colnames(numerical)


for (i in 1:length(Numerical))
{
  assign(paste0("gn",i), ggplot(aes_string(y = (Numerical[i]), x = "Churn"), data = subset(df))+ 
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="red", fill = "blue" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=Numerical[i],x="churn")+
           ggtitle(paste("Box plot for",Numerical[i])))
}

gridExtra::grid.arrange(gn1,gn2,gn3,ncol=3)
gridExtra::grid.arrange(gn4,gn5,gn6,ncol=3)
gridExtra::grid.arrange(gn7,gn8,gn9,ncol=3)
gridExtra::grid.arrange(gn10,gn11,gn12,ncol=3)
gridExtra::grid.arrange(gn13,gn14,gn15,ncol=3)
gridExtra::grid.arrange(gn16,ncol=1)





##Data Manupulation; convert string categories into factor numeric
#assigning levels to categorical variables
fnames = c("state","international.plan","voice.mail.plan","Churn")
for(i in fnames){
  if(class(df_train[,i])== 'factor'){
    df_train[,i] = factor(df_train[,i], labels = (1:length(levels(factor(df_train[,i])))))
  }}
for(i in fnames){
  if(class(df_test[,i])== 'factor'){
    df_test[,i] = factor(df_test[,i], labels = (1:length(levels(factor(df_test[,i])))))
  }
}
rm(fnames)
rm(i)
numeric_index = sapply(df_train,is.numeric) #selecting only numeric
numeric_data = df_train[,numeric_index]
cnames = colnames(numeric_data)
## Correlation Plot for numeric variables
corrgram(df_train[,numeric_index], order = F,
         upper.panel=panel.pie, text.panel=panel.txt, main = "Correlation Plot")
## Chi-squared Test of Independence for categorical variable
factor_index = sapply(df_train,is.factor)
factor_data = df_train[,factor_index]
for (i in 1:4)
{
  print(names(factor_data)[i])
  print(chisq.test(table(factor_data$Churn,factor_data[,i])))
}
## Dimension Reduction
df_train = subset(df_train, select = -c(total.day.minutes, total.eve.minutes, total.night.minutes, total.intl.minutes, phone.number,state))
df_test = subset(df_test, select = -c(total.day.minutes, total.eve.minutes, total.night.minutes, total.intl.minutes, phone.number,state))
#normality check
hist(df_train$number.customer.service.calls)
#Normalisation
cnames = c("account.length","area.code","number.vmail.messages","total.day.calls","total.day.charge",
           "total.eve.calls","total.eve.charge","total.night.calls","total.night.charge","total.intl.calls", "total.intl.charge",
           "number.customer.service.calls")
for(i in cnames){
  print(i)
  df_train[,i] = (df_train[,i] - min(df_train[,i]))/
    (max(df_train[,i] - min(df_train[,i])))
}
for(i in cnames){
  print(i)
  df_test[,i] = (df_test[,i] - min(df_test[,i]))/
    (max(df_test[,i] - min(df_test[,i])))
} 
rmExcept(c("df_train","df_test"))
train = df_train
test = df_test
##Decision tree for classification
#Develop Model on training data
C50_model = C5.0(Churn ~., train, trials = 50, rules = TRUE)
#Summary of DT model
summary(C50_model)
#write rules into disk
write(capture.output(summary(C50_model)), "c50Rules.txt")
#Lets predict for test cases
C50_Predictions = predict(C50_model, test[,-16], type = "class")
##Evaluate the performance of classification model
ConfMatrix_C50 = table(test$Churn, C50_Predictions)
confusionMatrix(ConfMatrix_C50)
#False Negative rate
FNR = FN/(FN+TP)
#Accuracy: 88.62%
#FNR: 7.4%
###Random Forest
RF_model = randomForest(Churn ~ ., train, importance = TRUE, ntree = 150)
#Extract rules fromn random forest
#transform rf object to an inTrees' format
treeList = RF2List(RF_model)
#
#Extract rules
exec = extractRules(treeList, train[,-16]) # R-executable conditions
#Visualize some rules
exec[1:2,]
#Make rules more readable:
readableRules = presentRules(exec, colnames(train))
readableRules[1:2,]
#Get rule metrics
ruleMetric = getRuleMetric(exec, train[,-16], train$Churn) # get rule metrics
#evaulate few rules
ruleMetric[1:2,]
#Presdict test data using random forest model
RF_Predictions = predict(RF_model, test[,-16])
##Evaluate the performance of classification model
ConfMatrix_RF = table(test$Churn, RF_Predictions)
confusionMatrix(ConfMatrix_RF)
#False Negative rate
FNR = FN/FN+TP
#Accuracy = 89.98
#FNR = 6.8%
#Logistic Regression
logit_model = glm(Churn ~ ., data = train, family = "binomial")
#summary of the model
summary(logit_model)
#predict using logistic regression
logit_Predictions = predict(logit_model, newdata = test, type = "response")
#convert prob
logit_Predictions = ifelse(logit_Predictions > 0.5, 1, 0)



##Evaluate the performance of classification model
ConfMatrix_LR = table(test$Churn, logit_Predictions)

#False Negative rate
FNR = FN/FN+TP
#Accuracy: 86.26 %
#FNR: 75.0 %
##KNN Implementation
library(class)
#Predict test data
KNN_Predictions = knn(train[, 1:15], test[, 1:15], train$Churn, k = 7)
#Confusion matrix
ConfMatrix_Knn = table(KNN_Predictions, test$Churn)
confusionMatrix(ConfMatrix_Knn)
#False Negative rate
FNR = FN/FN+TP
test
#Accuracy = 86.62
sum(diag(ConfMatrix_Knn))/nrow(test)
#FNR = 48.14
#naive Bayes
library(e1071)
#Develop model
NB_model = naiveBayes(Churn ~ ., data = train)
#predict on test cases #raw
NB_Predictions = predict(NB_model, test[,1:15], type = 'class')
#Look at confusion matrix
Confmatrix_NB = table(observed = test[,16], predicted = NB_Predictions)
confusionMatrix(Confmatrix_NB)
#Accuracy: 87.16
#FNR: 53.57

File: churn_reduction.py

# coding: utf-8

# # Python code for Churn Reduction

# In[ ]:

# Load libraries
import os
import pandas as pd
import numpy as np
from fancyimpute import KNN
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency
import seaborn as sns
from random import randrange, uniform


# In[2]:

#Setting path
os.chdir("E:/1ST SEM/eng/edwisor_assignments/8.project 2")


# # Loading data

# In[3]:

#loading train and test data
df_train = pd.read_csv("Train_data.csv")
df_test = pd.read_csv("Test_data.csv")


# In[4]:

cnames = ["account length","area code","number vmail messages","total day minutes","total day calls","total day charge",
"total eve minutes","total eve calls","total eve charge","total night minutes","total night calls",
"total night charge","total intl minutes","total intl calls", "total intl charge",
"number customer service calls"]


# # Outlier Analysis

# In[5]:

d1=df_train[cnames]
d2=df_test[cnames]


# In[ ]:




# In[6]:

for i in cnames:
    #Detect and replace with NA
    #Extract quartiles
    q75, q25 = np.percentile(d1.loc[:,i], [75 ,25])

    #Calculate IQR
    iqr = q75 - q25

    #Calculate inner and outer fence
    mini = q25 - (iqr*1.5)
    maxi = q75 + (iqr*1.5)

    #Replace with NA
    d1.loc[d1.loc[:,i] < mini,:i] = np.nan
    d1.loc[d1.loc[:,i] > maxi,:i] = np.nan
    # df_train["state"]=NULL
    #impute with median
    #d1[:,i]=d1[:,i].fillna(d1[:,i].median())


# In[7]:

for i in cnames:
    #Detect and replace with NA
    #Extract quartiles
    q75, q25 = np.percentile(d2.loc[:,i], [75 ,25])

    #Calculate IQR
    iqr = q75 - q25

    #Calculate inner and outer fence
    mini = q25 - (iqr*1.5)
    maxi = q75 + (iqr*1.5)

    #Replace with NA
    d2.loc[d2.loc[:,i] < mini,:i] = np.nan
    d2.loc[d2.loc[:,i] > maxi,:i] = np.nan
    # df_train["state"]=NULL
    #impute with median
    #d1[:,i]=d1[:,i].fillna(d1[:,i].median())


# ### MIssing value imputation after outlier analysis

# In[8]:

d1=d1.apply(lambda x:x.fillna(x.median()),axis=0)
d2=d2.apply(lambda x:x.fillna(x.median()),axis=0)


# In[9]:

d1["international plan"]=df_train["international plan"]
d1["voice mail plan"]=df_train["voice mail plan"]
d1["churn"]=df_train["Churn"]

d2["international plan"]=df_test["international plan"]
d2["voice mail plan"]=df_test["voice mail plan"]
d2["churn"]=df_test["Churn"]


# In[10]:

#assigning levels to categorical varibales
for i in range(0, d1.shape[1]):
    if(d1.iloc[:,i].dtypes == 'object'):
        d1.iloc[:,i] = pd.Categorical(d1.iloc[:,i])
        d1.iloc[:,i] = d1.iloc[:,i].cat.codes

for i in range(0, d2.shape[1]):
    if(d2.iloc[:,i].dtypes == 'object'):
        d2.iloc[:,i] = pd.Categorical(d2.iloc[:,i])
        d2.iloc[:,i] = d2.iloc[:,i].cat.codes

d2.head(10)


# In[11]:

#storing target variable
train_targets = d1.churn
test_targets = d2.churn


# In[12]:

#combining train and test data for data prepocessing
combined = d1.append(d2)

print(combined.shape, d1.shape, d2.shape)


# In[13]:

combined.head()


# ## correlation plot

# In[14]:

df_corr = combined.loc[:,cnames]
get_ipython().run_line_magic('matplotlib', 'inline')
#correlation analysis
#set height and width of plot
f , ax = plt.subplots(figsize = (15,12))
#generate correlation matrix
corr = df_corr.corr()
#Plot using seaborn library
sns.heatmap(corr, mask=np.zeros_like(corr, dtype=np.bool), cmap='rainbow',annot=True,
            square=True, ax=ax)


# In[15]:

cat_names = ["international plan","voice mail plan"]

#chi square test of independence
for i in cat_names:
    print(i)
    chi2, p, dof, ex = chi2_contingency(pd.crosstab(combined["churn"],combined[i]))
    print(p)


# In[16]:

combined.info()


# ## Dimensionality Reduction

# In[17]:

#dropping unnecessary variables
combined = combined.drop(["total day minutes", "total eve minutes", "total night minutes", "total intl minutes"], axis = 1)

combined.shape


# In[18]:

cnames = ["account length","area code","number vmail messages","total day calls","total day charge",
"total eve calls","total eve charge","total night calls","total night charge","total intl calls",
"total intl charge", "number customer service calls"]


# ## Normalization

# In[19]:

#normalization
for i in cnames:
    print(i)
    combined[i] = (combined[i]-min(combined[i]))/(max(combined[i])-min(combined[i]))

combined.head(10)


# In[20]:

#loading libraries for model
from sklearn import tree
from sklearn.metrics import accuracy_score
from sklearn.cross_validation import train_test_split
from sklearn.model_selection import cross_val_score
#splitting combined data to train and test
train = combined[:3333]

test = combined[3333:]


# In[21]:

features=["account length","area code","number vmail messages","total day calls","total day charge",
"total eve calls","total eve charge","total night calls","total night charge","total intl calls", "total intl charge",
"number customer service calls","international plan","voice mail plan"]


# In[22]:

X=combined[features]


# In[23]:

y=combined.churn


# ## Decision Tree model

# In[24]:

#decision tree model
c50_model = tree.DecisionTreeClassifier(criterion = 'entropy').fit(train, train_targets)
c50_pred = c50_model.predict(test)

c50_pred

#dot file to look at decision tree
dotfile = open("pt.dot", 'w')
df = tree.export_graphviz(c50_model, out_file=dotfile, feature_names = train.columns)

#testing accuracy of model
from sklearn.metrics import confusion_matrix
CM = pd.crosstab(test_targets, c50_pred)
CM

TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]
accuracy_score(test_targets,c50_pred)*100
#accuracy = 92.32153
#(FN*100)/(FN+TP)
#FNR = 32.142857
#(TP*100)/(TP+FN)
#Recall = 67.8571428


# ## Cross Validation

# In[25]:

# 10-fold cross-validation with logistic regression
print(cross_val_score(c50_model, X, y, cv=10, scoring='accuracy').mean())


# In[26]:

accuracy_score(test_targets,c50_pred)*100


# In[27]:

from sklearn.model_selection import cross_val_score


# ## Random Forest model

# In[28]:

#random forest model
from sklearn.ensemble import RandomForestClassifier
RF_model = RandomForestClassifier(n_estimators = 100).fit(train,train_targets)
RF_prediction = RF_model.predict(test)


RF_prediction

CM = pd.crosstab(test_targets, RF_prediction)
CM

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]
#check accuracy of model
accuracy_score(test_targets, RF_prediction)*100
#((TP+TN)*100)/(TP+TN+FP+FN)
#accuracy = 94.96100
#(FN*100)/(FN+TP)
#FNR = 33.928571


# In[29]:

# 10-fold cross-validation with K=5 for KNN (the n_neighbors parameter)
scores = cross_val_score(RF_model, X, y, cv=10, scoring='accuracy')
print(scores)


# ## KNN Model

# In[ ]:

#KNN implementation
from sklearn.neighbors import KNeighborsClassifier
KNN_model = KNeighborsClassifier(n_neighbors = 9).fit(train, train_targets)

#predict test cases
KNN_Predictions = KNN_model.predict(test)

#build confusion matrix
CM = pd.crosstab(test_targets, KNN_Predictions)
CM

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]
#check accuracy of model
#accuracy_score(y_test, y_pred)*100
((TP+TN)*100)/(TP+TN+FP+FN)
#accuracy = 86.622675
#False Negative rate
#(FN*100)/(FN+TP)
#FNR =97.321428



# In[31]:

from sklearn.model_selection import cross_val_score


# In[32]:

# 10-fold cross-validation with K=5 for KNN (the n_neighbors parameter)
scores = cross_val_score(KNN_model, X, y, cv=10, scoring='accuracy')
print(scores)


# ## Naive Bayes

# In[33]:

#Naive Bayes
from sklearn.naive_bayes import GaussianNB
#Naive Bayes implementation
NB_model = GaussianNB().fit(train, train_targets)

#predict test cases
NB_Predictions = NB_model.predict(test)

#Build confusion matrix
CM = pd.crosstab(test_targets, NB_Predictions)
CM

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]
#check accuracy of model
#accuracy_score(Y_test, y_pred)*100
((TP+TN)*100)/(TP+TN+FP+FN)
#Accuracy = 85.842831433
#False Negative rate
#(FN*100)/(FN+TP)
#FNR = 60.2678571

#we will be fixing random forest model as it provides best results
#now we will generate example out for out sample input test data with Random forest predictions
move = pd.DataFrame(RF_prediction)
move = move.rename(columns = {0:'move'})

test = test.join(move['move'])

test.to_csv("example_output.csv", index = False)


# In[ ]:

# 10-fold cross-validation with K=5 for KNN (the n_neighbors parameter)
scores = cross_val_score(NB_model, X, y, cv=10, scoring='accuracy')
print(scores)

